### ISSUE TITLE
Is it possible to tell linq2db to generate string literals instead of variables?

### ISSUE THREAD

**Issue by glebkozh** [linq2db/linq2db#1230]
Hello! I've got the following issue.
I am trying to use Sql Server 2016 JSON_VALUE functionality via linq2db's Sql.Expression attribute. I've created the following expression:

```cs
[Sql.Expression("JSON_VALUE(FileAttributesJsonColumn, {0})", ServerSideOnly = true)]
public static object JsonValue(string propertyName) => throw new InvalidOperationException();
```
Then I use it like that:
```cs
var files = fileTable.Where(_ => SqlExtensions.JsonValue("$.Extension") == "pdf").ToArray();`
```
In this case linq2db generates pretty valid where clause:

`JSON_VALUE(FileAttributesJson, N'$.Extension') = N'pdf'`

However when i try to use variable instead of string literal as property name I receive an error:

```cs
var value = "$.Extension";
var files = fileTable.Where(_ => SqlExtensions.JsonValue(value) == "pdf").ToArray();
```
`System.Data.SqlClient.SqlException : The argument 2 of the "JSON_VALUE or JSON_QUERY" must be a string literal.`

That happens because in this case linq2db generates variable:

```sql
DECLARE @value1 NVarChar(4000) -- String
SET     @value1 = N'$.Extension'
```
and uses it in where clause:

`JSON_VALUE(FileAttributesJson, @value1) = N'pdf'`

while Sql Server 2016 allows only literals as JSON_VALUE arguments. 

Is there a way to tell linq2db to inject C# variable value directly into SQL string literal instead of variable?
Thanks a lot!

**Comment by glebkozh** [comment]
@sdanyliv, awesome, thanks! And one more related question: can I pass FileAttributesJsonColumn as parameter? I mean, if I pass column name as string linq2db generates N prefix and quotes which is not desirable in case of column name.

**Comment by sdanyliv** [comment]
Almost everything is possible, but looks like in this situation you need `Sql.ExtensionAttribute` with custom builder. Will prepare sample.

**Comment by sdanyliv** [comment]
BTW this also should work: 
```cs
[Sql.Expression("JSON_VALUE({0}, {1})", ServerSideOnly = true, InlineParameters = true)]
public static string JsonValue(object expression, string path) => throw new InvalidOperationException();
```
Usage:
```cs
var files = fileTable
   .Where(_ => SqlExtensions.JsonValue(_.FileAttributesJsonColumn, "$.Extension") == "pdf")
   .ToArray();
```

**Comment by glebkozh** [comment]
@sdanyliv oh, it was really straightforward. Thanks a lot!

**Comment by CoskunSunali** [comment]
@sdanyliv Is there a way the above code could be converted to something like the following using the ExtensionAttribute? Or some other way?

```cs
    [Table(Name = "Content")]
    public class Content<TData>
        where TData: class, new()
    {
        public Content()
        {
        }

        [PrimaryKey, Identity]
        public int Id { get; set; }

        [Column, NotNull]
        public string Name { get; set; }

        [Column(DataType = DataType.NText), NotNull, SOMECUSTOMEXTENSIONATTRIBUTE]
        public TData Data { get; set; } // This is the JSON column

        [Column(SkipOnInsert = true, SkipOnUpdate = true)]
        public string PageTitle { get; set; }
    }

    // Query
    GetTable<Content<MyObject>>().Where(e => e.Data.SomeProperty == "Value");
```

And get the attribute to create the `JSON_VALUE([Data], '$.SomeProperty')` query automatically using the `SOMECUSTOMEXTENSIONATTRIBUTE` attribute assigned to the `Data` property?

**Comment by sdanyliv** [comment]
A lot of things are possible. Better to manipulate JSON object and real examples without abstraction. Only in this case i can help you.
We have member mapping, so you can map JObject’s property or method to your custom extension method.
http://blog.linq2db.com/#MapMember
It is like constructor.

**Comment by CoskunSunali** [comment]
Thank you for your reply!

> Better to manipulate JSON object and real examples without abstraction.

I am not sure if I understood your statement. The property is not of type JObject but just some custom class that is serialized using the following piece of code.

```cs
            MappingSchema.Default.SetConverter<MyObject, string>((obj) =>
            {
                if (obj == null)
                {
                    return null;
                }

                return Newtonsoft.Json.JsonConvert.SerializeObject(obj);
            });

            MappingSchema.Default.SetConverter<MyObject, DataParameter>((obj) =>
            {
                if (obj == null)
                {
                    return new DataParameter { DataType = LinqToDB.DataType.NText, Value = DBNull.Value };
                }

                return new DataParameter { Value = Newtonsoft.Json.JsonConvert.SerializeObject(obj) };
            });

            MappingSchema.Default.SetConverter<string, MyObject>((txt) =>
            {
                if (string.IsNullOrEmpty(txt))
                {
                    return null;
                }

                return Newtonsoft.Json.JsonConvert.DeserializeObject<MyObject>(txt);
            });
```

And what I am trying to avoid is to create a custom expression method as follows:

```cs
[Sql.Expression("JSON_VALUE({0}, {1})", ServerSideOnly = true, InlineParameters = true)]
public static string JsonValue(object expression, string path) => throw new InvalidOperationException();
```

Which results in other developers to be have to call the `JsonValue` method while querying against the `Data` column which is stored as json within the database.

```cs
var files = fileTable
   .Where(_ => SqlExtensions.JsonValue(_.FileAttributesJsonColumn, "$.Extension") == "pdf")
   .ToArray();
```

Actually, I am looking for a sample implementation of `ExtensionAttribute` where I can create the `JSON_VALUE({0}, {1})` by myself based on the `ISqlExpression` or something.

So basically when someone says `Where(e => e.Data.SomeProperty == "Value")`, I know that the `Data` is a json column and `SomeProperty` is a property on it which should result in `WHERE JSON_VALUE([Data], '$.SomeProperty') == 'Value')`.

Do you think that is likely with `ExtensionAttribute` or am I on the wrong way?

**Comment by sdanyliv** [comment]
I’m away from computer but will try to write sample:
```cs
LinqToDB.Linq.Expressions.MapMember(
    (JObject o) => o.SomeProperty,
    o => MySqlExtensions.JsonProperty(o.SomeProperty));
```
So when you will use SomeProperty in linq expression, linq2db will automatically convert this call to `JsonProperty` which is extension method that represents SQL.

It is more clear?

**Comment by CoskunSunali** [comment]
You are so helpful! It is definitely much clear.

May I ask why do you give examples based on `JObject`, is it because you think that I am using JObject or is it because that is how it has to be implemented? I am a bit confused.

My `e.Data` is of type, let's say `MyObject` and its structure is basically as follows, just to avoid any misunderstanding:

```cs
    public class MyObject
    {
        public string SomeProperty { get; set; }

        public string OtherProperty { get; set; }
    }
```

So you can assume the `Content` POCO as follows:

```cs
    [Table(Name = "Content")]
    public class Content
    {
        [PrimaryKey, Identity]
        public int Id { get; set; }

        [Column, NotNull]
        public string Name { get; set; }

        [Column(DataType = DataType.NText), NotNull]
        public MyObject Data { get; set; }
    }
```

And since you mentioned `MySqlExtensions.JsonProperty` method, is it not specific to MySQL? Can I find a similar method for MSSQL?

Sorry for my too many questions, hopefully I will soon get a better hold of linq2db and stop bothering you :)

**Comment by sdanyliv** [comment]
Sorry my fault, i mean `MySqlExtensions` as your static class, that you have defined somewhere.
Check our sources how we define a lot of extensions: 
https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Sql/Sql.cs

Also you can find in sources how we map a tons of members to these extensions.

**Comment by sdanyliv** [comment]
Or for example, how i have mapped EF function `Like` to our implementation.
https://github.com/linq2db/linq2db.EntityFrameworkCore/blob/master/Source/LinqToDB.EntityFrameworkCore/LinqToDBForEFTools.Mapping.cs#L18

**Comment by CoskunSunali** [comment]
There is not a single reason to be sorry. You are doing great and probably helping me in your personal time which I really appreciate.

I will go ahead and investigate the last two links you shared and see how you implemented things.

Very helpful, thank you a million times!

**Comment by CoskunSunali** [comment]
@sdanyliv Unfortunately I had no luck implementing a "dynamic" way of generating the JSON_VALUE function.

What I am ultimately trying to achieve is if I include the following piece of lambda expression in a where statement, to get it to be converted to a JSON_VALUE expression on the server side.

`Where(e => e.Data.Property1 == 'value')`

If I could only hook into the event where the SQL statements are being generated, I would check if the `Data` property has an attribute like `JsonColumn` and generate the rest of the expression myself, something similar to:

`JSON_VALUE(Data, '$.Property1')`

Of course I also need to handle the cases where there are inner properties:

`Where(e => e.Data.Property1.InnerProperty1 == 'value')` which I am trying to convert to something like:

`JSON_VALUE(Data, '$.Property1.InnerProperty1')`

I understand that it is my responsibility to generate the expression on the fly since there is no support for this out of the box (I guess there is for postgresql?) but just cannot figure out how to "get into" linq2db so that I can extend it.

Any assistance you might provide would be really appreciated if possible.

**Comment by sdanyliv** [comment]
@CoskunSunali there is closest solution that i can advise. Also you can handle Expression tree before converting SQL, but i have no time to play with that.
```cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using LinqToDB;
using LinqToDB.Data;
using LinqToDB.Expressions;
using LinqToDB.Extensions;
using LinqToDB.Mapping;
using LinqToDB.SqlQuery;
using Newtonsoft.Json;
using NUnit.Framework;

namespace Tests.Playground
{

	[AttributeUsage(AttributeTargets.Property)]
	public class JSonContentAttribute : Attribute
	{
	}

	public static class MappingHelper
	{
		private static MethodInfo _deserializeMethod;
		private static MethodInfo _serializeMethod;
		private static ConstructorInfo _dataParamContructor;

		static MappingHelper()
		{
			_deserializeMethod = MemberHelper.MethodOf(() => JsonConvert.DeserializeObject(null, typeof(int)));
			_serializeMethod = MemberHelper.MethodOf(() => JsonConvert.SerializeObject(null));
			_dataParamContructor = typeof(DataParameter).GetConstructor(new[] { typeof(string), typeof(object) });
		}

		public static void GenerateConvertorsForTables(Type dataConnectionType, MappingSchema ms)
		{
			var propsWithTables = dataConnectionType.GetProperties()
				.Where(p => typeof(IQueryable<>).IsSameOrParentOf(p.PropertyType));

			var types = propsWithTables.Select(p => p.PropertyType.GenericTypeArguments[0]).Distinct().ToArray();
			foreach (var t in types)
			{
				GenerateConvertors(t, ms);
			}
		}

		public static void GenerateConvertors(Type entityType, MappingSchema ms)
		{
			foreach (var propertyInfo in entityType.GetProperties().Where(p => p.GetCustomAttribute(typeof(JSonContentAttribute)) != null))
			{
				// emulating inParam => JsonConvert.DeserializeObject(inParam, propertyInfo.PropertyType)

				var inParam = Expression.Parameter(typeof(string));
				var lambda = Expression.Lambda(
					Expression.Convert(
						Expression.Call(null, _deserializeMethod, inParam,
							Expression.Constant(propertyInfo.PropertyType)),
						propertyInfo.PropertyType),
					inParam);

				ms.SetConvertExpression(typeof(string), propertyInfo.PropertyType, lambda);

				var inObjParam = Expression.Parameter(propertyInfo.PropertyType);
				var lambdaSql = Expression.Lambda(
					Expression.New(_dataParamContructor, Expression.Constant("p"),
						Expression.Call(null, _serializeMethod, inObjParam))
					, inObjParam);

				ms.SetConvertExpression(propertyInfo.PropertyType, typeof(DataParameter), lambdaSql);
			}
		}
	}

	public class MyDataConnection : DataConnection
	{
		private static MappingSchema _convertorSchema;

		static MyDataConnection()
		{
			_convertorSchema = new MappingSchema();
			MappingHelper.GenerateConvertorsForTables(typeof(MyDataConnection), _convertorSchema);
		}

		public MyDataConnection(string providerName, string connectionString, MappingSchema mappingSchema) : base(providerName, connectionString, mappingSchema)
		{
			AddMappingSchema(_convertorSchema);
		}

		public MyDataConnection(string configurationString) : base(configurationString)
		{
			AddMappingSchema(_convertorSchema);
		}

		public ITable<SampleClass> SampleClass => GetTable<SampleClass>();
	}

	[Table]
	public class SampleClass
	{
		[Column] public int Id    { get; set; }

		
		[Column(DataType = DataType.VarChar, Length = 4000), JSonContent]
		public DataClass Data { get; set; }
	}

	public class DataClass
	{
		public string Property1 { get; set; }
	}

	public static class Json
	{
		class JsonValueBuilder : Sql.IExtensionCallBuilder
		{
			public void Build(Sql.ISqExtensionBuilder builder)
			{
				var pathExpr = builder.Arguments[0];
				if (pathExpr.NodeType != ExpressionType.MemberAccess)
					throw new NotSupportedException();

				var pathList = new List<Expression>();
				var current = pathExpr;
				while (true)
				{
					pathList.Add(current);
					if (current.NodeType == ExpressionType.MemberAccess)
					{
						current = ((MemberExpression) current).Expression;
					}
					else
						break;
				}

				pathList.Reverse();

				var entity = pathList[0];
				var field  = pathList[1];

				var fieldSql = builder.ConvertExpressionToSql(field);
				builder.AddParameter("field", fieldSql);

				var propPathStr = "$";
				for (int i = 2; i < pathList.Count; i++)
				{
					propPathStr += "." + ((MemberExpression) pathList[i]).Member.Name;
				}

				builder.AddParameter("propPath", new SqlValue(propPathStr));
			}
		}


		[Sql.Extension("JSON_VALUE({field}, {propPath})", Precedence = Precedence.Primary, BuilderType = typeof(JsonValueBuilder), ServerSideOnly = true)]
		public static string Value(object path)
		{
			throw new NotImplementedException();
		}
	}

	[TestFixture]
	public class TestJson : TestBase
	{

		[Test, Combinatorial]
		public void SampleSelectTest([DataSources(false)] string context)
		{
			using (var db = new MyDataConnection(context))
			using (var table = db.CreateLocalTable<SampleClass>())
			{
				db.Insert(new SampleClass { Id = 1, Data = new DataClass { Property1 = "Pr1" } });

				var objects = table.Where(t => Sql.AsNotNull(Json.Value(t.Data.Property1)) == "Pr1")
					.ToArray();
			}
		}

	}
}
```

**Comment by sdanyliv** [comment]
@CoskunSunali, did my solution help in your situation? 
Yes there are many code, but for introduction some generalized method we have to do it carefully. BTW, i think it should be done in 3.0.

**Comment by CoskunSunali** [comment]
@sdanyliv First of all, sorry for the late reply.

That being said, you have no idea how much your code helped! It was really inspiring and I successfully achieved the first important step of generalizing the data access. I made some adjustments here and there, since my `MyDataConnection` does not have those `Table<>` properties but anyway, those are just details to the implementation. 

I also started to better understand Linq2DB internals, e.g. what does your `MappingSchema` does.

May I ask, is there any reason why I would like to create a new static MappingSchema instance instead of using `MappingSchema.Default`?

What is more, currently the following works great:

```cs
var objects = table.Where(t => Sql.AsNotNull(Json.Value(t.Data.Property1)) == "Pr1")
					.ToArray();
```

However, I would like to wrap the `Where` extension method and modify the expression right before it is passed to the SQL generators. 

So I would like to, myself, add the `Json.Value` call expression by checking the provided where statement dynamically without requiring my developers to be have to write `Json.Value(t.Data.Property1) == "Pr1"` and instead just write `t.Data.Property1 == "Pr1"`.

Is there an event that I can hook into, a callback I can provide somewhere or an interface I can implement for being able to extend Linq2DB in such a way or what would be your approach for that if you were to do so?

Finally, please please please do not spend your personal and valuable time to write the whole code again. I really appreciate your assistance but at the same time aware of that you are doing it in your personal time and I am sure you have more important things to do then writing code for every question I ask you :)

Thanks a million times!

**Comment by sdanyliv** [comment]
There is `IExpressionPreprocessor` interface. Implement it in your `DataConnection` class.
You can use our extension `Transform` which simplifies things a lot when modifying Expression Tree.

You can setup `MappingSchema.Default` - then converters will be available globally.

**Comment by CoskunSunali** [comment]
Thanks for your reply! I will go ahead and play with that interface and the `Transform` class. I will let you know if I am stuck somewhere.

By the way, the above code you shared with me which helped me a lot now contains SQL statements specific to MSSQL Server. How may I make sure that it also works for MySQL? Not asking you to write the MySQL query statement, just wondering if I can improve the code to handle databases that support JSON data type other than MSSQL too.

Consider that MSSQL function for JSON value is `MSSQL_JSON_VALUE` and for MySQL is `MySQL_JSON_VALUE`, would that be as the following?

```cs
[Sql.Extension("MSSQL_JSON_VALUE({field}, {propPath})", Precedence = Precedence.Primary, BuilderType = typeof(JsonValueBuilder), ServerSideOnly = true)]
		public static string MSSQLValue(object path)
		{
			throw new NotImplementedException();
		}

[Sql.Extension("MySQL_JSON_VALUE({field}, {propPath})", Precedence = Precedence.Primary, BuilderType = typeof(JsonValueBuilder), ServerSideOnly = true)]
		public static string MySQLValue(object path)
		{
			throw new NotImplementedException();
		}
```

In short, would I be writing two separate `Sql.Extension` methods, is that the right approach?

Thank you yet another time.

**Comment by sdanyliv** [comment]
>Finally, please please please do not spend your personal and valuable time to write the whole code again. I really appreciate your assistance but at the same time aware of that you are doing it in your personal time and I am sure you have more important things to do then writing code for every question I ask you :)

I had to. Otherwise i will spend a lot of much more time explaining. Sample is usually better :)
Also i have never did created such converters before, so it was my personal experience and i saw new ways how to improve this area.

**Comment by sdanyliv** [comment]
Check how it is done in this file. Also here you can find a lot of undocumented extensions.
https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Sql/Sql.cs#L208

**Comment by CoskunSunali** [comment]
Ahh, the `Sql.Extension` attribute takes a `builderType` parameter! Great!

I have to admit that the documentation sucks :( but you are probably well aware of it.

Thanks a lot again. Hope to be able to handle the rest regarding JSON data myself. Cheers!

**Comment by sdanyliv** [comment]
Documentation is a pain. We have to drop everything off and write articles.
Also check this blog: http://blog.linq2db.com/
